msal/imds.py

```diff
@@ -0,0 +1,164 @@
+# Copyright (c) Microsoft Corporation.
+# All rights reserved.
+#
+# This code is licensed under the MIT License.
+import json
+import logging
+import os
+import socket
+import time
+try:  # Python 2
+    from urlparse import urlparse
+except:  # Python 3
+    from urllib.parse import urlparse
+
+logger = logging.getLogger(__name__)
+
+def _scope_to_resource(scope):  # This is an experimental reasonable-effort approach
+    u = urlparse(scope)
+    if u.scheme:
+        return "{}://{}".format(u.scheme, u.netloc)
+    return scope  # There is no much else we can do here
+
+
+def _obtain_token(http_client, resource, client_id=None):
+    if "IDENTITY_ENDPOINT" in os.environ and "IDENTITY_HEADER" in os.environ:
+        return _obtain_token_on_app_service(
+            http_client, os.environ["IDENTITY_ENDPOINT"], os.environ["IDENTITY_HEADER"],
+            resource, client_id=client_id)
+    return _obtain_token_on_azure_vm(http_client, resource, client_id=client_id)
+
+
+def _obtain_token_on_azure_vm(http_client, resource, client_id=None):
```

Comment:
Besides `client_id`, `object_id` and `msi_res_id` should also be supported. Azure CLI heavily relies on all 3 of them: https://github.com/Azure/azure-cli/blob/060b414d8ac5e4d8f4fc17cc2e23c2154c762226/src/azure-cli-core/azure/cli/core/_profile.py#L195-L227

They are documented at - https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http (Also see https://github.com/MicrosoftDocs/azure-docs/issues/15288#issuecomment-619531505)

`msrestazure`'s implementation: https://github.com/Azure/msrestazure-for-python/blob/8e3def8c9927a349596e575de93df76de7a3a2b6/msrestazure/azure_active_directory.py#L468

Reply:
Different Managed Identity services support different IDs. `client_id` seems to be the only common ID.

@jiasli , do you know some scenarios that the customer would have to use a different-than-`client_id` id?

Reply:
> do you know some scenarios that the customer would have to use a different-than-`client_id` id?

No, but since the managed identity endpoint supports all 3 of them, we need to do the same. Dropping `object_id` and `msi_res_id` will be a breaking change.

Reply:
> Besides client_id, object_id and msi_res_id should also be supported. [Azure CLI heavily relies on all 3 of them](https://github.com/Azure/azure-cli/blob/060b414d8ac5e4d8f4fc17cc2e23c2154c762226/src/azure-cli-core/azure/cli/core/_profile.py#L195-L227)

On a close look, Azure CLI does not surface an "identity id type" concept, therefore the usage pattern is based on inefficient trial-and-error. Don't you consider either adding the identity id type, or [tentatively deprecate](https://docs.python.org/3/library/exceptions.html#PendingDeprecationWarning) the object_id and msi_res_id and see how many users are actually using it?

Reply:
The cost of this "inefficient trial-and-error" is pretty low. It only occurs during `az login --identity`. Later operations will know the type of the ID and directly specify the type.

Unfortunately, we don't have telemetry for monitoring the usage of `object_id` and `msi_res_id`.

Reply:
The latest commit added support for `object_id` and `msi_res_id`.